internal/locate/region_cache.go

```diff
@@ -508,6 +508,132 @@ func (mu *regionIndexMu) refresh(r []*Region) {
 
 type livenessFunc func(ctx context.Context, s *Store) livenessState
 
+// repeat wraps a `func()` as a schedulable fuction for `bgRunner`.
+func repeat(f func()) func(context.Context, time.Time) bool {
+	return func(_ context.Context, _ time.Time) bool {
+		f()
+		return false
+	}
+}
+
+// until wraps a `func() bool` as a schedulable fuction for `bgRunner`.
+func until(f func() bool) func(context.Context, time.Time) bool {
+	return func(_ context.Context, _ time.Time) bool {
+		return f()
+	}
+}
+
+type bgRunner struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func newBackgroundRunner(ctx context.Context) *bgRunner {
+	ctx, cancel := context.WithCancel(ctx)
+	return &bgRunner{
+		ctx:    ctx,
+		cancel: cancel,
+	}
+}
+
+func (r *bgRunner) closed() bool {
+	select {
+	case <-r.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (r *bgRunner) shutdown(wait bool) {
+	r.cancel()
+	if wait {
+		r.wg.Wait()
+	}
+}
+
+// run calls `f` once in background.
+func (r *bgRunner) run(f func(context.Context)) {
+	if r.closed() {
+		return
+	}
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		f(r.ctx)
+	}()
+}
+
+// schedule calls `f` every `interval`.
+func (r *bgRunner) schedule(f func(context.Context, time.Time) bool, interval time.Duration) {
+	if r.closed() {
+		return
+	}
+	ticker := time.NewTicker(interval)
```

Comment:
How about moving this line after line 575? since the ticker is stop in another goroutine.

Reply:
My point is: if the interval <= 0, it panic when calling schedule, which is able to be recovered by the caller; however if we move it into the goroutine, panic cannot be caught by users.

Reply:
If so, we should check interval <= 0 here, instead of causing panic and recover to handle it?

Reply:
IMO, panic on interval <= 0 is reasonable, but the panic should be recoverable. I will update it as you said anyway.

Reply:
I think we should try to check in advance to avoid panic, Instead of letting it panic, after all, this is a system software, users will panic when they see panic logs.

Reply:
I've changed it.
BTW, panic let you kown you've passed a wrong arg, however, keeping silent sometimes hide potential risks.